pkg/provider/azure_controller_vmss.go

```diff
@@ -189,21 +191,45 @@ func (ss *ScaleSet) DetachDisk(ctx context.Context, nodeName types.NodeName, dis
 			},
 		},
 	}
-	// Invalidate the cache right after updating
+
+	updateResult := compute.VirtualMachineScaleSetVM{}
+	var rerr *retry.Error
+
 	defer func() {
-		_ = ss.DeleteCacheForNode(vmName)
+		// If there is an error with Update operation,
+		// invalidate the cache
+		if rerr != nil {
+			_ = ss.DeleteCacheForNode(vmName)
+			return
+		}
+
+		// Update the cache with the updated result
+		vmssVMCacheEntry := &vmssVirtualMachinesEntry{
```

Comment:
Consider passing the parameters required to initialize the `vmssVirtualMachinesEntry` struct to the `updateCache` function and initializing it there instead. This would simplify the deferred functions and reduce redundant code.

Reply:
Fixed